elastic-buffer-sim/src/Bittide/Simulate.hs

```diff
@@ -84,65 +85,118 @@ tunableClockGen settlePeriod periodOffset stepSize _reset speedChange =
     in
       newPeriod :- go newSettleCounter newPeriod scs
 
--- | Determines how 'elasticBuffer' should respond to underflow/overflow.
-data OverflowMode
-  -- | Saturate at empty/full boundaries. Useful for testing purposes.
-  = Saturate
-  -- | Error on write-when-full, or read-when-empty. This mode shuold be used in
-  -- practise. Overflowing elastic buffer cannot happen in a real Bittide system.
-  | Error
+type ResetClockControl = Bool
 
--- | Simple model of a FIFO that only models the interesting part for conversion:
--- data counts.
+-- | This wrapper disables reads or writes when the elastic buffer
+-- over-/underflows, as appropriate.
+--
+-- It also censors 'DataCount' after an over-/underflow, so that we do not take
+-- measurements from an elastic buffer until it has settled back to its
+-- midpoint.
+ebController ::
+  forall readDom writeDom.
+  (KnownDomain readDom, KnownDomain writeDom) =>
+  ElasticBufferSize ->
+  Clock readDom ->
+  Clock writeDom ->
+  (Signal readDom DataCount, Signal readDom ResetClockControl)
+ebController size clkRead clkWrite =
+  unbundle
+    (go <$> rdToggle <*> outRd <*> overflowRd)
+ where
+  (outRd, outWr) = elasticBuffer size clkRead clkWrite rdToggle wrToggle
+
+  go True (dc, False) False = (dc, False)
+  go _ _ _ = (deepErrorX "Data count censored", True)
+
+  overflowRd =
+    dualFlipFlopSynchronizer clkWrite clkRead resetGen enableGen False (snd <$> outWr)
+
+  rdToggle =
+    register clkRead resetGen enableGen True
+      $ mealy clkRead resetGen enableGen f False outRd
+   where
+    f :: Bool -> (DataCount, Underflow) -> (Bool, Bool)
```

Comment:
I can't keep up with all the `Bool`s, could you add type aliases + a comment on what this is doing?